### PR TITLE
Add better merge extension

### PR DESCRIPTION
**Pull request by pprice**
Adds the better merge extension to vscode proper. (https://marketplace.visualstudio.com/items?itemName=pprice.better-merge). This will enable decoration, code-lenses and commands for "standard" style merge conflicts. 

Notable changes:
 - `better-merge.*` command contributions changed to `merge-conflict.*` with same keybindings and functionality. 
 - Settings collapsed down to `merge-conflict.codeLens.enabled` and `merge-conflict.decorators.enabled`
 - Replace the single pass regex scanning of entire document, with line-by-line scanning, sans regex. 
 - Added localization (`package.nls.json` and `vcode-nls` usage) support
 - Remove the `git` extension decoration of merge blocks. 

Known TODOs: 
 - Performance measurements for large files. Maximum to test up to is 5mb as anything larger is not synchronized with extensions. 
 - It is possible to fire two scan **consumptions** from the same "origin" code, this mostly happens in `mergeDecorator.ts`, when a document is opened then active. The actual scan happens once as it is debounced via the delayer, BUT the merge decorators are sent to the document twice. I have a candidate fix in the works. 
 - Replace hard-coded decorator colors with something to get theme colors when available. 

Potential TODOs (probably future PRs) 
 - Support diff3 format

### CODE REVIEW

File: extensions/merge-conflict/package.json [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        "key": "alt+m 2"
+      },
+      {
+        "command": "merge-conflict.accept.both",
+        "when": "editorTextFocus",
+        "key": "alt+m 3"
+      }
+    ],
+    "configuration": {
+      "title": "Merge Conflict",
```
Comment (chrmarti) [inline extensions/merge-conflict/package.json]: Extract to nls file for completeness.

File: extensions/merge-conflict/src/commandHandler.ts [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+		});
+
+		const leftTitle = `Current changes`; // (Ln ${range.start.line}${!range.isSingleLine ? `-${range.end.line}` : ''})`;
+
+		range = conflict.incoming.content;
+		const rightUri = leftUri.with({ query: JSON.stringify(range) });
+
+		const rightTitle = `Incoming changes`; // (Ln${range.start.line}${!range.isSingleLine ? `-${range.end.line}` : ''})`;
+
+		const title = `${fileName}: ${leftTitle} \u2194 ${rightTitle}`;
```
Comment (chrmarti) [inline extensions/merge-conflict/src/commandHandler.ts]: Maybe inline the two titles and use a placeholder `{0}` to localize like: https://github.com/Microsoft/vscode/blame/master/extensions/git/src/commands.ts#L754

File: extensions/merge-conflict/package.json (hunk 1 above)
Comment (chrmarti) [inline extensions/merge-conflict/package.json]: Extract to nls file for completeness.

File: extensions/merge-conflict/src/documentTracker.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+			this.cache.delete(key);
+		}
+	}
+
+	dispose() {
+		this.cache.clear();
+	}
+
+	private getConflictsOrEmpty(document: vscode.TextDocument): interfaces.IDocumentMergeConflict[] {
+		let stepStart = process.hrtime();
```
Comment (chrmarti) [inline extensions/merge-conflict/src/documentTracker.ts]: Skip or remove console output.

File: extensions/merge-conflict/package.json [hunk 4]
```diff
@@ -0,0 +1,128 @@
+{
+  "name": "merge-conflict",
+  "publisher": "vscode",
+  "displayName": "merge-conflict",
+  "description": "Merge Conflict",
+  "version": "0.7.0",
+  "aiKey": "AIF-d9b70cd4-b9f9-4d70-929b-a071c400b217",
+  "enableProposedApi": true,
```
Comment (chrmarti) [inline extensions/merge-conflict/package.json]: Not needed, I believe. Remove the property to clarify.

File: extensions/merge-conflict/src/typings/refs.d.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+/// <reference path='../../../../src/vs/vscode.d.ts'/>
+/// <reference path='../../../../src/vs/vscode.proposed.d.ts'/>
```
Comment (chrmarti) [inline extensions/merge-conflict/src/typings/refs.d.ts]: Not needed, I believe. Remove to clarify.

**Comment by niyazpk** [comment]
Just noticed this feature during a merge conflict. This is just great. Thanks a bunch!!!

**Pull request merged**